### PR TITLE
Improve robustness of ISO initrd builds

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

The ISO target (`make iso ...`) can be fragile, both to changes in the files included in the ISO's initrd, and to parallel builds. We can harden this by tracking the files included with the ISO in the makefile, as well as defining some order-only prerequisites for the initrd target.

Order-only prerequisites mean that a given target will allways be built after another target, but changing the prerequisite won't actually case a follow-on rebuild of the target. (ie `a: b | c`, `c` will always be completed before `a` starts, but changing `c` won't cause `a` to rebuild during future builds). By making all of the `iso:` target's prerequisites also order-only prerequisites for the `$(initrd_img)` target, we can guarantee that all the build actions needed to prep for `iso:` will be completed before we make the recursive call to build the initrd. (The makefile was trying to build the worker chroot twice in parallel in the same directory, this did not end well)

###### Change Log  

- Make `iso:`'s prerequisites order-only prerequisites for `$(initrd_img)` as well to avoid parallel builds during the recursive call
- Track all the assets we include in `$(initrd_img)` so that changes to those files will always cause a rebuild of the initrd.

###### Does this affect the toolchain?  

**NO**

###### Associated issues  

- https://microsoft.visualstudio.com/OS/_workitems/edit/42963350
- https://microsoft.visualstudio.com/OS/_workitems/edit/42951530/

###### Test Methodology

- Local ISO & image builds

### CODE REVIEW

File: toolkit/scripts/imggen.mk [hunk 1]
```diff
@@ -163,11 +166,17 @@ $(image_external_package_cache_summary): $(cached_file) $(go-imagepkgfetcher) $(
 		--output-summary-file=$@ \
 		--output-dir=$(external_rpm_cache)
 
-$(initrd_img): $(go-liveinstaller) $(go-imager) $(initrd_config_json) $(INITRD_CACHE_SUMMARY)
+# We need to ensure that initrd_img recursive build will never run concurrently with another build component, so add all ISO prereqs as 
+#	order-only-prerequisites to initrd_img
```
Comment (MateuszMalisz) [inline toolkit/scripts/imggen.mk:170]: nit:tab

File: toolkit/scripts/imggen.mk [hunk 2]
```diff
@@ -16,6 +16,10 @@ initrd_config_json       = $(RESOURCES_DIR)/imageconfigs/iso_initrd_arm64.json
 else
 initrd_config_json       = $(RESOURCES_DIR)/imageconfigs/iso_initrd.json
 endif
+initrd_assets_dir        = $(RESOURCES_DIR)/imageconfigs/additionalfiles/iso_initrd/
+initrd_scripts_dir       = $(RESOURCES_DIR)/imageconfigs/postinstallscripts/iso_initrd/
+initrd_packages_dir      = $(RESOURCES_DIR)/imageconfigs/packagelists
+initrd_assets_files      = $(shell find $(initrd_assets_dir) $(initrd_scripts_dir) $(initrd_packages_dir))
```
Comment (PawelWMS) [inline toolkit/scripts/imggen.mk:22]: Could we introduce this small change to be more precise about the dependencies?

```suggestion
initrd_packages_json     = $(RESOURCES_DIR)/imageconfigs/packagelists/iso-initrd-packages.json
initrd_assets_files      = $(initrd_packages_json) $(shell find $(initrd_assets_dir) $(initrd_scripts_dir))
```

**Review by PawelWMS — APPROVED** [review]
I love the change. I've added just one small comment to consider.

File: toolkit/scripts/imggen.mk (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/scripts/imggen.mk]: Sure, I'll need to track all three files though

Comment (dmcilvaney) [reply toolkit/scripts/imggen.mk]: https://github.com/microsoft/CBL-Mariner/pull/4601/commits/e08c6f87961c902bea41320000ad2f2079e3b9c1

**Pull request merged**